File: web/src/core/detection/calibration.ts
/**
 * Calibration — port of detection/calibration.py. See docs/CALIBRATION.md
 * for the full rationale, including two correctness fixes carried over
 * from the Python side (do not regress these here):
 *   1. Filter-rebound candidates are excluded from calibration stats, and
 *      only the top-N-by-amplitude candidate(s) per trial are kept (one for
 *      SINGLE_BLINK, two for DOUBLE_BLINK) — otherwise residual filter
 *      ringing gets pooled into the stats and drags the median down.
 *   2. min_relative_prominence_spread defaults to 0.5 (50% of the
 *      calibrated median), not a tight value — a double blink's second
 *      pulse measures a legitimately lower prominence than an isolated
 *      blink (it rides on the first pulse's still-decaying filter tail),
 *      and a tight floor derived from a low-variance calibration session
 *      can reject it.
 *   3. The single-blink-derived floor above is still only a guess. When
 *      DOUBLE_BLINK trial data exists, deriveConfigOverrides also computes
 *      a hard ceiling from the *actually measured* second-pulse
 *      prominence/duration (real hardware bug, 2026-09-15: a calibration
 *      that passed 3/3 single + 3/3 double still rejected 15/16 live
 *      candidates afterwards) — see deriveConfigOverrides' own comments.
 */
import type { CalibrationStats } from "../types";
import { defaultCalibrationStats } from "../types";
import type { BciConfig, DeepPartial } from "../config";
import { MAD_TO_SIGMA, median, medianAbsoluteDeviation } from "./adaptiveThreshold";
import { CandidateDetector } from "./candidateDetector";
import { extractFeatures } from "./features";
import type { BlinkCandidate } from "../types";

function medianMad(values: number[]): [number, number] {
  if (values.length === 0) return [0, 0];
  const med = median(values);
  return [med, medianAbsoluteDeviation(values, med)];
}

export interface TrialSegment {
  label: "REST" | "SINGLE_BLINK" | "DOUBLE_BLINK";
  frontal: Float64Array;
  af7: Float64Array;
  af8: Float64Array;
}

function peakUv(c: BlinkCandidate): number {
  let m = 0;
  for (let i = 0; i < c.frontalWindow.length; i++) m = Math.max(m, Math.abs(c.frontalWindow[i]));
  return m;
}

function findCandidatesInSegment(seg: TrialSegment, fsHz: number): BlinkCandidate[] {
  const detector = new CandidateDetector({ fsHz, thresholdWindowS: Math.max(seg.frontal.length / fsHz, 1.0) });
  const out: BlinkCandidate[] = [];
  for (let i = 0; i < seg.frontal.length; i++) {
    const cand = detector.processSample(seg.frontal[i], seg.af7[i], seg.af8[i]);
    if (cand !== null && cand.widthValid && !cand.likelyFilterRebound) out.push(cand);
  }
  return out;
}

export function computeCalibrationStats(segments: TrialSegment[], fsHz: number): CalibrationStats {
  const stats = defaultCalibrationStats();

  const restValues: number[] = [];
  for (const seg of segments) {
    if (seg.label === "REST") {
      for (let i = 0; i < seg.frontal.length; i++) restValues.push(Math.abs(seg.frontal[i]));
      stats.nRestTrials += 1;
    } else if (seg.label === "SINGLE_BLINK") {
      stats.nSingleTrials += 1;
    } else if (seg.label === "DOUBLE_BLINK") {
      stats.nDoubleTrials += 1;
    }
  }
  [stats.baselineMedian, stats.baselineMad] = medianMad(restValues);
  stats.noiseFloorMedian = stats.baselineMedian;
  stats.noiseFloorMad = stats.baselineMad;

  const normalPeaks: number[] = [];
  const intentionalPeaks: number[] = [];
  const intentionalDurations: number[] = [];
  const intentionalRiseTimes: number[] = [];
  const intentionalFallTimes: number[] = [];
  const doubleSpacings: number[] = [];
  const doubleSecondPulseProminences: number[] = [];
  const doubleSecondPulseDurations: number[] = [];

  for (const seg of segments) {
    const candidates = findCandidatesInSegment(seg, fsHz);
    if (seg.label === "REST") {
      for (const c of candidates) normalPeaks.push(peakUv(c));
    } else if (seg.label === "SINGLE_BLINK") {
      if (candidates.length > 0) {
        const c = candidates.reduce((best, cur) => (peakUv(cur) > peakUv(best) ? cur : best));
        intentionalPeaks.push(peakUv(c));
        intentionalDurations.push(c.durationS);
        let peakIdx = 0;
        let bestAbs = -Infinity;
        for (let i = 0; i < c.frontalWindow.length; i++) {
          const v = Math.abs(c.frontalWindow[i]);
          if (v > bestAbs) {
            bestAbs = v;
            peakIdx = i;
          }
        }
        intentionalRiseTimes.push(peakIdx / fsHz);
        intentionalFallTimes.push((c.frontalWindow.length - 1 - peakIdx) / fsHz);
      }
    } else if (seg.label === "DOUBLE_BLINK") {
      if (candidates.length >= 2) {
        const topTwo = [...candidates].sort((a, b) => peakUv(b) - peakUv(a)).slice(0, 2);
        const [a, b] = topTwo.sort((x, y) => x.startTimeS - y.startTimeS);
        const spacing = b.startTimeS - a.endTimeS;
        if (spacing > 0) doubleSpacings.push(spacing);
        // b is the second (later) pulse — the one riding the first pulse's
        // still-decaying filter tail. Measure its *actual*
        // classifier-equivalent prominence/duration (not just peak
        // amplitude) so deriveConfigOverrides can use real data, not a
        // guess, as the ceiling on minProminenceUv/minBlinkWidthS.
        const bFeatures = extractFeatures(
          b.frontalWindow, b.af7Window, b.af8Window, fsHz,
          stats.noiseFloorMedian, spacing > 0 ? spacing : 0,
        );
        doubleSecondPulseProminences.push(bFeatures.peakProminence);
        doubleSecondPulseDurations.push(b.durationS);
      }
    }
  }

  [stats.normalBlinkPeakMedian, stats.normalBlinkPeakMad] = medianMad(normalPeaks);
  stats.nRestCandidates = normalPeaks.length;

  [stats.intentionalBlinkPeakMedian, stats.intentionalBlinkPeakMad] = medianMad(intentionalPeaks);
  [stats.intentionalDurationMedianS, stats.intentionalDurationMadS] = medianMad(intentionalDurations);
  [stats.intentionalRiseTimeMedianS] = medianMad(intentionalRiseTimes);
  [stats.intentionalFallTimeMedianS] = medianMad(intentionalFallTimes);
  stats.nSingleCandidates = intentionalPeaks.length;

  [stats.doubleBlinkSpacingMedianS, stats.doubleBlinkSpacingMadS] = medianMad(doubleSpacings);
  stats.nDoublePairs = doubleSpacings.length;

  [stats.doubleBlinkSecondPulseProminenceMedian, stats.doubleBlinkSecondPulseProminenceMad] =
    medianMad(doubleSecondPulseProminences);
  [stats.doubleBlinkSecondPulseDurationMedianS, stats.doubleBlinkSecondPulseDurationMadS] =
    medianMad(doubleSecondPulseDurations);

  return stats;
}

export function deriveConfigOverrides(
  stats: CalibrationStats,
  prominenceMarginSigma = 1.5,
  intervalMarginS = 0.15,
  minRelativeProminenceSpread = 0.5,
  minDurationSpreadS = 0.02,
  secondPulseMarginSigma = 1.0,
  minRelativeSecondPulseSpread = 0.3,
): DeepPartial<BciConfig> & { calibrationStats: CalibrationStats } {
  const prominenceSpread = Math.max(
    stats.intentionalBlinkPeakMad * MAD_TO_SIGMA,
    minRelativeProminenceSpread * stats.intentionalBlinkPeakMedian,
  );
  let minProminence = Math.max(
    stats.intentionalBlinkPeakMedian - prominenceMarginSigma * prominenceSpread,
    stats.noiseFloorMedian + 2 * MAD_TO_SIGMA * stats.noiseFloorMad,
    1e-6,
  );

  // The above is still only a *guess* at how much weaker a second pulse
  // will be, extrapolated from single-blink amplitude. When DOUBLE_BLINK
  // trial data is available, doubleBlinkSecondPulseProminenceMedian is the
  // real measured peakProminence (extractFeatures, the exact feature the
  // live classifier gates on) of those trials' second pulses — used here
  // as a hard ceiling so the derived threshold can never exceed what was
  // actually measured. Deliberately NOT re-clamped to the noise-floor lower
  // bound above: these second pulses were already real, width-valid,
  // non-artifact candidates during calibration, so re-imposing that bound
  // here would recreate the exact bug this ceiling exists to fix.
  if (stats.nDoublePairs > 0 && stats.doubleBlinkSecondPulseProminenceMedian > 0) {
    const secondPulseSpread = Math.max(
      stats.doubleBlinkSecondPulseProminenceMad * MAD_TO_SIGMA,
      minRelativeSecondPulseSpread * stats.doubleBlinkSecondPulseProminenceMedian,
    );
    const secondPulseCeiling = Math.max(
      stats.doubleBlinkSecondPulseProminenceMedian - secondPulseMarginSigma * secondPulseSpread,
      1e-6,
    );
    minProminence = Math.min(minProminence, secondPulseCeiling);
  }

  const durationSpread = Math.max(stats.intentionalDurationMadS, minDurationSpreadS);
  let minWidth = Math.max(stats.intentionalDurationMedianS - 3 * durationSpread, 0.02);
  const maxWidth = stats.intentionalDurationMedianS + 4 * durationSpread + 0.1;

  if (stats.nDoublePairs > 0 && stats.doubleBlinkSecondPulseDurationMedianS > 0) {
    // Same reasoning as the prominence ceiling: a fast double blink's
    // second-pulse candidate window can be genuinely shorter than an
    // isolated blink's — never derive a minBlinkWidthS above what was
    // actually measured.
    const secondPulseDurationSpread = Math.max(stats.doubleBlinkSecondPulseDurationMadS, minDurationSpreadS);
    minWidth = Math.min(
      minWidth,
      Math.max(stats.doubleBlinkSecondPulseDurationMedianS - 2 * secondPulseDurationSpread, 0.02),
    );
  }

  const overrides: DeepPartial<BciConfig> & { calibrationStats: CalibrationStats } = {
    candidateDetection: {
      minProminenceUv: round(minProminence, 4),
      minBlinkWidthS: round(minWidth, 4),
      maxBlinkWidthS: round(maxWidth, 4),
    },
    calibrationStats: stats,
  };

  if (stats.nDoublePairs > 0) {
    const spacing = stats.doubleBlinkSpacingMedianS;
    const spacingSpread = Math.max(stats.doubleBlinkSpacingMadS * MAD_TO_SIGMA, 0.05);
    const minInterval = Math.max(spacing - spacingSpread - intervalMarginS, 0.08);
    const maxInterval = spacing + spacingSpread + intervalMarginS;
    overrides.doubleBlink = { minIntervalS: round(minInterval, 4), maxIntervalS: round(maxInterval, 4) };
  }

  return overrides;
}

function round(value: number, digits: number): number {
  const factor = 10 ** digits;
  return Math.round(value * factor) / factor;
}

File: src/bcihand/detection/calibration.py
"""Calibration: turn a short guided recording session into per-user thresholds.

We deliberately do NOT ship one universal fixed amplitude threshold. Instead
we run a short guided sequence (REST / REST / REST / SINGLE / SINGLE / SINGLE
/ DOUBLE / DOUBLE / DOUBLE, per docs/CALIBRATION.md) and derive robust
(median / MAD) statistics from it.

Honest limitation (see docs/SIGNAL_PIPELINE.md "Known limitations"):
physiologically, a deliberate single blink and a spontaneous single blink
often look very similar in amplitude/shape. Calibration cannot manufacture a
separation that isn't there. That is *why* this system's primary control
signal is the double blink (an unusual, low-base-rate temporal pattern)
rather than the single blink — see config/default_config.yaml
`communication.command_mapping`.
"""
from __future__ import annotations

from dataclasses import dataclass, field
from pathlib import Path

import numpy as np
import yaml

from .adaptive_threshold import MAD_TO_SIGMA
from .candidate_detector import CandidateDetector, BlinkCandidate


def _median_mad(values: list[float]) -> tuple[float, float]:
    if not values:
        return 0.0, 0.0
    arr = np.asarray(values, dtype=np.float64)
    med = float(np.median(arr))
    mad = float(np.median(np.abs(arr - med)))
    return med, mad


@dataclass
class CalibrationStats:
    baseline_median: float = 0.0
    baseline_mad: float = 0.0
    noise_floor_median: float = 0.0
    noise_floor_mad: float = 0.0

    normal_blink_peak_median: float = 0.0
    normal_blink_peak_mad: float = 0.0
    n_rest_candidates: int = 0

    intentional_blink_peak_median: float = 0.0
    intentional_blink_peak_mad: float = 0.0
    intentional_duration_median_s: float = 0.0
    intentional_duration_mad_s: float = 0.0
    intentional_rise_time_median_s: float = 0.0
    intentional_fall_time_median_s: float = 0.0
    n_single_candidates: int = 0

    double_blink_spacing_median_s: float = 0.0
    double_blink_spacing_mad_s: float = 0.0
    n_double_pairs: int = 0

    n_rest_trials: int = 0
    n_single_trials: int = 0
    n_double_trials: int = 0

    def derive_config_overrides(self, prominence_margin_sigma: float = 1.5,
                                 interval_margin_s: float = 0.15,
                                 min_relative_prominence_spread: float = 0.5,
                                 min_duration_spread_s: float = 0.02) -> dict:
        """Turn measured stats into a config-override dict compatible with
        config/default_config.yaml's structure (deep-merged as
        config/calibration_active.yaml).

        A short calibration session (a handful of trials, or trials that
        happen to be unusually consistent — e.g. our own simulator's
        deterministic pulses) can produce a measured MAD of ~0. Without a
        floor, that collapses min_prominence_uv to essentially the exact
        calibration-trial amplitude and min_blink_width_s to its exact
        duration, rejecting perfectly genuine blinks that are only slightly
        weaker/shorter than the calibration sample. `min_relative_prominence_spread`
        and `min_duration_spread_s` are floors on the spread used, applied
        symmetrically with the existing max_blink_width_s floor below.

        min_relative_prominence_spread defaults to a generous 0.5 (50% of the
        calibrated single-blink median) rather than a tight value, because
        the SINGLE_BLINK calibration trials are isolated blinks, but a
        double blink's *second* pulse rides on the first pulse's still-decaying
        causal-filter tail and its measured peak_prominence (which subtracts
        deviation from baseline — see detection/features.py) is legitimately
        lower than an isolated blink's, often by 40-50%. This was found by
        direct observation: with a tighter floor, scripts/calibrate.py
        --source simulate followed by scripts/run_pipeline.py silently
        turned both scripted demo double blinks into unmatched SINGLE
        blinks (the second pulse's prominence fell just under
        min_prominence_uv). See also the calibrate.py post-calibration
        self-check, which replays the calibration trials themselves through
        the full classifier and warns if a derived threshold would reject
        one of its own double-blink second pulses.
        """
        prominence_spread = max(
            self.intentional_blink_peak_mad * MAD_TO_SIGMA,
            min_relative_prominence_spread * self.intentional_blink_peak_median,
        )
        min_prominence = max(
            self.intentional_blink_peak_median - prominence_margin_sigma * prominence_spread,
            self.noise_floor_median + 2 * MAD_TO_SIGMA * self.noise_floor_mad,
            1e-6,
        )

        duration_spread = max(self.intentional_duration_mad_s, min_duration_spread_s)
        min_width = max(self.intentional_duration_median_s - 3 * duration_spread, 0.02)
        max_width = self.intentional_duration_median_s + 4 * duration_spread + 0.1

        if self.n_double_pairs > 0:
            spacing = self.double_blink_spacing_median_s
            spacing_spread = max(self.double_blink_spacing_mad_s * MAD_TO_SIGMA, 0.05)
            min_interval = max(spacing - spacing_spread - interval_margin_s, 0.08)
            max_interval = spacing + spacing_spread + interval_margin_s
        else:
            min_interval, max_interval = None, None

        overrides: dict = {
            "candidate_detection": {
                "min_prominence_uv": round(min_prominence, 4),
                "min_blink_width_s": round(min_width, 4),
                "max_blink_width_s": round(max_width, 4),
            },
            "calibration_stats": self.__dict__,
        }
        if min_interval is not None:
            overrides["double_blink"] = {
                "min_interval_s": round(min_interval, 4),
                "max_interval_s": round(max_interval, 4),
            }
        return overrides

    def save(self, path: Path | str) -> None:
        overrides = self.derive_config_overrides()
        with open(path, "w", encoding="utf-8") as f:
            yaml.safe_dump(overrides, f, sort_keys=False)


@dataclass
class TrialSegment:
    label: str  # "REST", "SINGLE_BLINK", "DOUBLE_BLINK"
    frontal: np.ndarray
    af7: np.ndarray
    af8: np.ndarray


def _find_candidates_in_segment(seg: TrialSegment, fs_hz: float) -> list[BlinkCandidate]:
    """Width-valid, non-rebound candidates within one trial segment.

    Filter-rebound lobes are excluded here for the same reason the live
    classifier hard-rejects them (see classification/blink_classifier.py):
    they are the causal bandpass filter's own opposite-sign overshoot
    following a real blink, not a second physiological event, and must never
    be pooled into calibration statistics as if they were.
    """
    detector = CandidateDetector(fs_hz=fs_hz, threshold_window_s=max(len(seg.frontal) / fs_hz, 1.0))
    out = []
    for f, a7, a8 in zip(seg.frontal, seg.af7, seg.af8):
        cand = detector.process_sample(f, a7, a8)
        if cand is not None and cand.width_valid and not cand.likely_filter_rebound:
            out.append(cand)
    return out


def _peak_uv(c: BlinkCandidate) -> float:
    return float(np.max(np.abs(c.frontal_window)))


def compute_calibration_stats(segments: list[TrialSegment], fs_hz: float) -> CalibrationStats:
    """segments: already-filtered trial recordings in calibration-sequence order."""
    stats = CalibrationStats()

    rest_values: list[float] = []
    for seg in segments:
        if seg.label == "REST":
            rest_values.extend(np.abs(seg.frontal).tolist())
            stats.n_rest_trials += 1
        elif seg.label == "SINGLE_BLINK":
            stats.n_single_trials += 1
        elif seg.label == "DOUBLE_BLINK":
            stats.n_double_trials += 1

    stats.baseline_median, stats.baseline_mad = _median_mad(rest_values)
    stats.noise_floor_median, stats.noise_floor_mad = stats.baseline_median, stats.baseline_mad

    normal_peaks: list[float] = []
    intentional_peaks: list[float] = []
    intentional_durations: list[float] = []
    intentional_rise_times: list[float] = []
    intentional_fall_times: list[float] = []
    double_spacings: list[float] = []

    for seg in segments:
        candidates = _find_candidates_in_segment(seg, fs_hz)
        if seg.label == "REST":
            for c in candidates:
                normal_peaks.append(float(np.max(np.abs(c.frontal_window))))
        elif seg.label == "SINGLE_BLINK":
            # A SINGLE_BLINK trial is, by the calibration protocol, exactly
            # one instructed blink — take only the largest-amplitude
            # candidate so any residual small same-sign filter ringing that
            # slips past the rebound guard (see docs/SIGNAL_PIPELINE.md
            # "Filter rebound rejection") cannot masquerade as extra blinks
            # and drag the median amplitude/duration estimate down.
            if candidates:
                c = max(candidates, key=_peak_uv)
                intentional_peaks.append(_peak_uv(c))
                intentional_durations.append(c.duration_s)
                peak_idx = int(np.argmax(np.abs(c.frontal_window)))
                intentional_rise_times.append(peak_idx / fs_hz)
                intentional_fall_times.append((len(c.frontal_window) - 1 - peak_idx) / fs_hz)
        elif seg.label == "DOUBLE_BLINK":
            # Likewise, a DOUBLE_BLINK trial is exactly two instructed
            # blinks: take the two largest-amplitude candidates (not simply
            # the first two in time) before computing their spacing.
            if len(candidates) >= 2:
                top_two = sorted(candidates, key=_peak_uv, reverse=True)[:2]
                a, b = sorted(top_two, key=lambda c: c.start_time_s)
                spacing = b.start_time_s - a.end_time_s
                if spacing > 0:
                    double_spacings.append(spacing)

    stats.normal_blink_peak_median, stats.normal_blink_peak_mad = _median_mad(normal_peaks)
    stats.n_rest_candidates = len(normal_peaks)

    stats.intentional_blink_peak_median, stats.intentional_blink_peak_mad = _median_mad(intentional_peaks)
    stats.intentional_duration_median_s, stats.intentional_duration_mad_s = _median_mad(intentional_durations)
    stats.intentional_rise_time_median_s, _ = _median_mad(intentional_rise_times)
    stats.intentional_fall_time_median_s, _ = _median_mad(intentional_fall_times)
    stats.n_single_candidates = len(intentional_peaks)

    stats.double_blink_spacing_median_s, stats.double_blink_spacing_mad_s = _median_mad(double_spacings)
    stats.n_double_pairs = len(double_spacings)

    return stats
